src/inmanta/ast/statements/__init__.py

```diff
@@ -230,6 +221,19 @@ def execute_direct(self, requires: abc.Mapping[str, object]) -> object:
         """
         raise DirectExecuteException(self, f"The statement {str(self)} can not be executed in this context")
 
+    def requires_emit_gradual(
+        self, resolver: Resolver, queue: QueueScheduler, resultcollector: ResultCollector[object]
+    ) -> dict[object, VariableABC]:
+        """
+        Returns a dict of the result variables required for execution. Behaves like requires_emit, but additionally may attach
+        resultcollector as a listener to result variables.
+        When this method is called, the caller must make sure to eventually call `execute` as well.
+
+        Composite statements (e.g. conditional expression) will pass result collectors to their children rather than
+        report to them themselves.
+        """
+        return {**self.requires_emit(resolver, queue), (self, ResultCollector): WrappedValueVariable(resultcollector)}
```

Comment:
I ended up passing it through the requires dict after all because it's a well-established method and I wanted to stick to one optimization at a time. I believe the other approach should be made sufficiently generic so that it can be used for e.g. the promises as well, and when I started on it I realized that it would take quite some time. I thought it better to defer it and create a ticket for it.